### PR TITLE
[Enhancement] - Improve regex in re-evaluate

### PR DESCRIPTION
**Pull request by FuchsDominik**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context

The regex in the short answer question edit component did not handle spots correctly which are directly connected to another word. For example
 - `[-spot 10];System.out.println();` was not split after `]`
 - `[index + 7]` was treated as spot
 - `myArray.[-spot 4]` was not split before `[`
This caused adjacent strings to disappear in the visual mode, which re-evaluate uses.

### Description

We improved the regex, which now handles also these situations

### Steps for Testing

1. Log in to Artemis
2. Create a short answer question exercise
3. Remove the whitespace between a spot and the following word
4. participate
5. click on re-evaluate
6. make sure the word following the spot is correctly displayed in the view

### Screenshot
For example a question like 

> [-spot 20].out.prinltn();

Results in:
Screenshot 2021-03-11 at 18 16 12

### CODE REVIEW

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 1]
```diff
@@ -74,6 +74,16 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
...
 
+    /**
+     * This is the regex used for splitting the text into words and spots.
+     * The regex splits the text the same was as {@link ShortAnswerQuestionUtil#divideQuestionTextIntoTextParts} does.
+     * The regex matches following cases:
+     * 1. All whitespace characters outside of spots
+     * 2. The left boundary of a spot if it is not immediately followed by a whitespace character
+     * 3. The right boundary of a spot if it is not immediately preceded by a whitespace character
+     */
+    regex = /(?<!\[-spot\s*)\s+|(?<=\[-spot\s*\d+\])(?=\S)|(?<=\S)(?=\[-spot\s*\d+\])/g;
```
Comment (krusche) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:85]: use a more descriptive name

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 2]
```diff
@@ -89,15 +99,15 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
...
-         * these [ ] brackets.
-         * **/
+         * 2. Now each line of the question text will be divided into each word (we use whitespace and the borders of spots as separator, see {@link #regex}).
+         */
         const textForEachLine = this.question.text!.split(/\n+/g);
         this.textParts = textForEachLine.map((t) => {
             const indentation = this.shortAnswerQuestionUtil.getIndentation(t);
-            const lineParts = t.split(/\s+(?![^[]?[\d]]*])/g);
-            lineParts[1] = indentation.concat(lineParts[1]);
+            const lineParts = t.split(this.regex);
```
Comment (krusche) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:107]: avoid one letter variables such as `t`

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 3]
```diff
@@ -616,7 +626,7 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
         this.question.spots = cloneDeep(this.backupQuestion.spots);
         // split on every whitespace. !!!only exception: [-spot 1] is not split!!! for more details see description in ngOnInit.
         const textForEachLine = this.question.text!.split(/\n+/g);
-        this.textParts = textForEachLine.map((t) => t.split(/\s+(?![^[]?[\d]]*])/g));
+        this.textParts = textForEachLine.map((t) => t.split(this.regex));
```
Comment (krusche) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:629]: avoid t

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 4]
```diff
@@ -662,7 +672,7 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
 
         // split on every whitespace. !!!only exception: [-spot 1] is not split!!! for more details see description in ngOnInit.
         const textForEachLine = this.question.text!.split(/\n+/g);
-        this.textParts = textForEachLine.map((t) => t.split(/\s+(?![^[]?[\d]]*])/g));
+        this.textParts = textForEachLine.map((t) => t.split(this.regex));
```
Comment (krusche) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:675]: avoid t

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 5]
```diff
@@ -692,7 +702,7 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
         this.question.text = this.textParts.map((textPart) => textPart.join(' ')).join('\n');
         // split on every whitespace. !!!only exception: [-spot 1] is not split!!! for more details see description in ngOnInit.
         const textForEachLine = this.question.text.split(/\n+/g);
-        this.textParts = textForEachLine.map((t) => t.split(/\s+(?![^[]?[\d]]*])/g));
+        this.textParts = textForEachLine.map((t) => t.split(this.regex));
```
Comment (krusche) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:705]: avoid t

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts (hunk 1 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts]: Used the term wordParserRegex in 84bdf3e

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts (hunk 2 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts]: 👍  84bdf3e

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts (hunk 3 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts]: 👍  84bdf3e

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts (hunk 4 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts]: 👍  84bdf3e

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts (hunk 5 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts]: 👍  84bdf3e

**Review by SimonEntholzer — APPROVED** [review]
Testing worked fine, although I think the visuals of the quiz-re-evaluate site could be improved, but that is not part of this PR.


I'm not really a pro when it comes to regexes, but as far as I understand it, it does make sense. I also double checked on [https://regexr.com/](https://regexr.com/), seems to do what it should.

**Pull request merged**